walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,54 +15,127 @@
+        // ranges that work for best for the mandatory attendees.
+        // If there are no mandatory attendees, treat the optional attendees as mandatory attendees
+        // and optimize the timeranges
+        if (request.getAttendees().size() == 0){
+            Collection<TimeRange> meetingTimeRangesWithOptionalAttendees = 
+                convertStateToTimeRanges(meetingStatesWithAllAttendees, duration, 
+                request.getAllAttendees().size(), true);
+            return meetingTimeRangesWithOptionalAttendees;
+        }else if (meetingTimeRangesWithAllAttendees.size() == 0) {
+            int[] meetingStatesWithMandatoryAttendees = 
+                getMinutesState(events, request.getAttendees());
+            Collection<TimeRange> meetingTimeRangesWithMandatoryAttendees = 
+                convertStateToTimeRanges(meetingStatesWithMandatoryAttendees, duration, 
+                request.getAttendees().size(), true);
+            return meetingTimeRangesWithMandatoryAttendees;
+        }else{
+            return meetingTimeRangesWithAllAttendees;
+        }
+    }
+
+    // Approach: assign numbers to every minute in the day representing their state
+    // 0 means not occupied, -1 means 1 person is not available at that minute, etc.
+    public int[] getMinutesState(Collection<Event> events, Collection<String> requestAttendees) {
+        int[] minutesState = new int[24 * 60]; 
+        for (Event event: events) {
+            for (int i = event.getWhen().start(); i < event.getWhen().end(); i++) {
+                minutesState[i] -= getNumOfAttendees(event, requestAttendees);
             }
-            // Add the last event if applicable.
-            if (TimeRange.END_OF_DAY - lastEnd >= duration){
-                validTimeRanges.add(TimeRange.fromStartEnd(lastEnd,TimeRange.END_OF_DAY,true));
+        }
+        return minutesState;
+    }
+
+    // Convert the array of states to a list of valid time ranges. If optimize is false, must return timeranges 
+    // that work for everyone (or an empty list if none); if optimize is true, return the best solution.
+    public Collection<TimeRange> convertStateToTimeRanges(int[] minutesState, 
+                long duration, long numOfAttendees, boolean optimize) {
+        Collection<TimeRange> meetingTimeRanges = Arrays.asList();
+        int max = Arrays.stream(minutesState).max().getAsInt(); 
+        int min = Arrays.stream(minutesState).min().getAsInt();
+        int j = 0;
+
+        if (!optimize){
+            return getTimeRangesBasedOnState(minutesState, duration, 0);
+        }
+
+        // The valid timeranges are sequences with consecutive best-state's.
+        for (j = max; j >= min; j--) {
+            // If we end up having to search the worst possible state, this means no one 
+            // is actually availble, so just return an empty list.
+            if (Math.abs(j) == numOfAttendees && numOfAttendees != 0) {
```

Comment:
Change |min| to `max(Arrays.stream(minutesState).min().getAsInt(), 1 - numOfAttendees)` and you can remove this if block.

Reply:
Done!